### PR TITLE
Load docstrings lazily from TASTy

### LINKED ISSUE #20106: Symbol.docString not available in tests
Using [Symbol.docstring](https://dotty.epfl.ch/api/scala/quoted/Quotes$reflectModule$SymbolMethods.html#docstring-fffffb50) from tests to inspect types defined in main always returns `None`.

## Compiler version

3.3.3
3.4.1

## Minimized code

https://github.com/OndrejSpanel/Scala3-DocStringInTest

### main:

```Scala
import scala.quoted.*

object Doc {
  inline def of[A]: Option[String] = ${ ofImpl[A] }

  def ofImpl[A: Type](using Quotes): Expr[Option[String]] = {
    import quotes.reflect.*

    val symbol = TypeRepr.of[A].typeSymbol
    Expr(symbol.docstring)
  }
}
```

### test:

```Scala
/**
 * Doc of Test
 * */
class Test

object Test {
  def main(args: Array[String]): Unit = {
    val docString = Doc.of[Main]
    println(docString)
    val docStringTest = Doc.of[Test]
    println(docStringTest)

    if (docString.isEmpty || docStringTest.isEmpty) System.exit(1)
  }
}
```

Use `Test/run` to run the code.

## Output

[Workflow output](https://github.com/OndrejSpanel/Scala3-DocStringInTest/actions/runs/8571450844/job/23491700317):

```scala
None
Some(/**
 * Doc of Test
 * */)
```

Note: while test can access values of `docstring` defined in test. it cannot access any `docstring` defined in main, it always gets `None`.

## Expectation

The `docstring` values of symbols defined in the main should be available from tests.

### PR DESCRIPTION
**Pull request by nicolasstucki**
Now we are always able to load docstrings from TASTy, even if `-Yread-docs` is not set. The `-Yread-docs` flag loads the doc strings eagerly, as it did before.

Fixes #20106

### CODE REVIEW

**Review by odersky — CHANGES_REQUESTED** [review]
When do we actually read comments? My concern is that creates a huge map in all cases now. We carefully engineered the Tasty unpickler to be fast and not create too much memory. Now we are throwing that out of the window for something as trivial as reading comments. I believe if we do that we have to first optimize comment handling as thoroughly as the rest of the representations. Right mow it is a completely different standard, nobody cared very much about efficiency, since we believed that in high-stress scenarios (clean build of a large code base) comments would not be read anyway. 

Unrelatedd: I have a similar concern for unused warnings which do something relatively "easy" yet need a significant part of the compilation time, simply because they have not been optimized well. 

Alternatives: Can we create/populate the map only once the first comment is read? Can we read the Tasty buffer twice to get comments? Or maybe have a more compact representation of where comments start, so that comments can be easily read on demand afterwards? Note that using a util.IntMap does not allocate any value objects. Maybe one could use that to get the comment start of a symbol.

File: scaladoc/src/dotty/tools/scaladoc/snippets/SnippetCompiler.scala [hunk 1]
```diff
@@ -31,7 +31,7 @@ class SnippetCompiler(
       rootCtx.setSetting(rootCtx.settings.experimental, true)
       rootCtx.setSetting(rootCtx.settings.YretainTrees, true)
       rootCtx.setSetting(rootCtx.settings.YcookComments, true)
-      rootCtx.setSetting(rootCtx.settings.YreadComments, true)
+      rootCtx.setSetting(rootCtx.settings.YreadDocsEagerly, true)
```
Comment (odersky) [inline scaladoc/src/dotty/tools/scaladoc/snippets/SnippetCompiler.scala:34]: What is a usage scenario where I might want readDocsEargerly?

**Review by odersky — CHANGES_REQUESTED** [review]
I think lazy loading is nice, but then

 - it has to be so fast that we don't need to worry about performance
 - it should avoid all settings, including readDocsEagerly.

File: scaladoc/src/dotty/tools/scaladoc/snippets/SnippetCompiler.scala (hunk 1 above)
Comment (nicolasstucki) [reply scaladoc/src/dotty/tools/scaladoc/snippets/SnippetCompiler.scala:34]: None. I just kept it for backward compat and for someone else to be able to benchmark the difference in performance.

**Comment by nicolasstucki** [comment]
We need to address the concerns and explore the alternatives described in  https://github.com/scala/scala3/pull/20148#pullrequestreview-1999552461

**Pull request closed without merging**